firedrake/matrix_free/preconditioners.py

```diff
@@ -0,0 +1,314 @@
+from __future__ import absolute_import
+import abc
+
+from firedrake.petsc import PETSc
+
+__all__ = ("AssembledPC", "MassInvPC", "PCDPC", "PCBase")
+
+
+class PCBase(object):
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self):
+        """Create a PC context suitable for PETSc.
+
+        Matrix free preconditioners should inherit from this class and
+        implement:
+
+        - :meth:`initialize`
+        - :meth:`update`
+        - :meth:`apply`
+        - :meth:`applyTranspose`
+
+        """
+        self.initialized = False
+        super(PCBase, self).__init__()
+
+    @abc.abstractmethod
+    def update(self, pc):
+        """Update any state in this preconditioner."""
+        pass
+
+    @abc.abstractmethod
+    def initialize(self, pc):
+        """Initialize any state in this preconditioner."""
+        pass
+
+    def setUp(self, pc):
+        """Setup method called by PETSc.
+
+        Subclasses should probably not override this and instead
+        implement :meth:`update` and :meth:`initialize`."""
+        if self.initialized:
+            self.update(pc)
+        else:
+            self.initialize(pc)
+            self.initialized = True
+
+    def view(self, pc, viewer=None):
+        if viewer is None:
+            return
+        typ = viewer.getType()
+        if typ != PETSc.Viewer.Type.ASCII:
+            return
+        viewer.printfASCII("Firedrake matrix-free preconditioner %s\n" %
+                           type(self).__name__)
+
+    @abc.abstractmethod
+    def apply(self, pc, X, Y):
+        """Apply the preconditioner to X, putting the result in Y.
+
+        Both X and Y are PETSc Vecs, Y is not guaranteed to be zero on entry.
+        """
+        pass
+
+    @abc.abstractmethod
+    def applyTranspose(self, pc, X, Y):
+        """Apply the transpose of the preconditioner to X, putting the result in Y.
+
+        Both X and Y are PETSc Vecs, Y is not guaranteed to be zero on entry.
+
+        """
+        pass
+
+
+class AssembledPC(PCBase):
+    """A matrix-free PC that assembles the operator.
+
+    Internally this makes a PETSc PC object that can be controlled by
+    options using the extra options prefix ``assembled_``.
+    """
+    def initialize(self, pc):
+        from firedrake.assemble import assemble
+
+        _, P = pc.getOperators()
+
+        context = P.getPythonContext()
+
+        # It only makes sense to preconditioner/invert a diagonal
+        # block in general.  That's all we're going to allow.
+        if not context.on_diag:
+            raise ValueError("Only makes sense to invert diagonal block")
+
+        self.P = assemble(context.a, bcs=context.row_bcs,
+                          form_compiler_parameters=context.fc_params, nest=False)
```

Comment:
Is this necessary?  My thought was that if we're assembling to get an algebraic PC (amg/ilu/direct method), that assembling a nest would not be needed.